torch/csrc/PyInterpreter.cpp

```diff
@@ -565,6 +565,149 @@ static void set_tensor_attr_with_capsule(
   py::handle(mb_obj.value()).attr(attr_name) = capsule;
 }
 
+// Note [Tensor Subclass custom size/stride caching strategy]
+// Tensor subclasses can use __torch_dispatch__ to override size/stride calls.
+// However, this presents a problem:
+// (1) When you return a custom (maybe symbolic) size/stride
+//     from python, we need to stash this fresh vector of ints/symints
+//     somewhere so that it has the same lifetime as the tensor.
+// (2) If the subclass experiences a metadata mutation,
+//     this stashed vector is no longer valid, so we need to allocate a fresh
+//     buffer to store the new sizes the next time someone asks for them.
+//
+// We handle this in the same way that `TensorImpl::sizes_default()`
+// handles its buffer: we simply reallocate the buffer whenever
+// the number of dimensions changes due to a resize.
+// Notable, we do *not* reallocate the buffer if the values changed,
+// but the number of dimensions stayed the same (e.g. `.transpose_()`).
+template <typename T>
+static c10::ArrayRef<T> get_set_cached_attr(
+    const c10::TensorImpl* tensor,
+    const char* base_attr_name,
+    const py::object& obj) {
+  c10::optional<PyObject*> mb_obj =
+      tensor->pyobj_slot()->check_pyobj(getPyInterpreter());
+  TORCH_CHECK(
+      mb_obj.has_value(), "Tensor subclass's PyInterpreter has no value");
+  auto buffer_len_attr_name = std::string(base_attr_name) + std::string("_len");
+  auto buffer_issmallvector_attr_name =
+      std::string(base_attr_name) + std::string("_issmallvector");
+
+  bool is_buffer_allocated = false;
+  bool is_smallvector = false;
+  size_t curr_size = 0;
+  if (PyObject_HasAttrString(mb_obj.value(), buffer_len_attr_name.c_str())) {
+    auto buffer_pyobj =
+        py::handle(mb_obj.value()).attr(buffer_len_attr_name.c_str());
+    curr_size = py::cast<size_t>(buffer_pyobj);
+    is_buffer_allocated = true;
+    TORCH_INTERNAL_ASSERT(PyObject_HasAttrString(
+        mb_obj.value(), buffer_issmallvector_attr_name.c_str()));
+    (PyObject_HasAttrString(
+        mb_obj.value(), buffer_issmallvector_attr_name.c_str()));
+    auto issmallvector_pyobj =
+        py::handle(mb_obj.value()).attr(buffer_issmallvector_attr_name.c_str());
+    is_smallvector = py::cast<bool>(issmallvector_pyobj);
+  }
+
+  size_t new_size = py::len(obj);
+
+  // It turns out that we need to maintain full-fidelity compared to
+  // TensorImpl::sizes() around when our buffer gets reallocated. In particular,
+  // SmallVector starts out by allocating a size-5 buffer. Any resizes on the
+  // tensor that are <= 5 elements will not reallocate the buffer. But once we
+  // hit a size that is > 5 elements, we will re-allocate on every future
+  // resize. NOTE: Ideally, we shouldn't need to model the SmallVector
```

Comment:
This doesn't seem quite accurate. If you go back down to <= 5 elements, you'll go back into the small vector regime. There isn't any hysteresis (every future resize)

Reply:
Ah whoops, I didn't read the SmallVector code carefully enough. Thanks, will update.